src/main/scala/com/databricks/sparkdl/ImageUtils.scala

```diff
@@ -0,0 +1,142 @@
+        channels match {
+          case 1 =>
+            b = imageData(offset)
+            g = b
+            r = b
+          case 3 =>
+            b = imageData(offset)
+            g = imageData(offset + 1)
+            r = imageData(offset + 2)
+          case _ =>
+            require(false, s"`Channels` must be 1 or 3, got $channels.")
+        }
+
+        val color = new Color(r & 0xff, g & 0xff, b & 0xff)
+        image.setRGB(w, h, color.getRGB)
+        offset += channels
+        w += 1
+      }
+      h += 1
+    }
+    image
+  }
+
+
+  /**
+   * Takes a Java BufferedImage and returns a Row Image (spImage). The
+   *
+   * @param image Java BufferedImage.
+   * @return Row image in spark.ml.image format with 3 channels in BGR order.
+   */
+  private[sparkdl] def spImageFromBufferedImage(image: BufferedImage): Row = {
+    val channels = 3
+    val height = image.getHeight
+    val width = image.getWidth
+
+    val decoded = new Array[Byte](height * width * channels)
+    var offset, h = 0
+    while (h < height) {
+      var w = 0
+      while (w < width) {
+        val color = new Color(image.getRGB(w, h))
+        decoded(offset) = color.getBlue.toByte
+        decoded(offset + 1) = color.getGreen.toByte
+        decoded(offset + 2) = color.getRed.toByte
+        offset += channels
+        w += 1
+      }
+      h += 1
+    }
+    // TODO: udpate mode to be Int when spark.ml.image is merged.
+    Row(null, height, width, channels, "CV_U8C3", decoded)
+  }
+
+  /**
+   * Resizes and image and returns it as an Array[Byte]. Only 1 and 3 channel inputs, where each
+   * channel is a single Byte, are currently supported. Only BGR channel order is supporte but
+   * this might work for other channel orders.
+   *
+   * @param tgtHeight   desired height of output image.
+   * @param tgtWidth    desired width of output image.
+   * @param tgtChannels number of channels of output image (must be 3), may be used later to
+   *                    support more channels.
+   * @param spImage     image to resize.
+   * @return resized image, if the input was BGR or 1 channel, the output will be BGR.
+   */
+  private[sparkdl] def resizeImage(
+    tgtHeight: Int,
+    tgtWidth: Int,
+    tgtChannels: Int)(spImage: Row): Row = {
+    require(tgtChannels == 3, s"`tgtChannels` was set to $tgtChannels, must be 3.")
+
+    val height = ImageSchema.getHeight(spImage)
+    val width = ImageSchema.getWidth(spImage)
+    val nChannels = ImageSchema.getNChannels(spImage)
+
+    if ((nChannels == tgtChannels) && (height == tgtHeight) && (width == tgtWidth)) {
+      spImage
+    } else {
+      val srcImg = spImageToBufferedImage(spImage)
+      val tgtImg = new BufferedImage(tgtWidth, tgtHeight, BufferedImage.TYPE_3BYTE_BGR)
+
+      // Draw srcImage onto resized (usually smaller) tgtImg
+      val graphic = tgtImg.createGraphics()
+      val scaledImg = srcImg.getScaledInstance(tgtWidth, tgtHeight, Image.SCALE_AREA_AVERAGING)
```

Comment:
The contract, as I believe, is that `DeepImageFeaturizer` should behave the same whether we call it from Scala or from Python. Let's find out a way to make sure they do and add tests.
Thank you!

Reply:
I think in the long term we'll make sure that the resizing is the same by having the python code call out to the scala implementation. Because of that I think we should focus on correctness here. I'm not sure what scaling method we're using in the python code, but we should use the AREA scaling method in both python and in scala, http://tanbakuchi.com/posts/comparison-of-openv-interpolation-algorithms/#Downsampling-comparison.

Reply:
Would this be solved by having the python version of `DeepImageFeaturizer` call into the scala version? If that's simple, we can do it as part of the `DeepImageFeaturizer` persistence work. There is no persistence support right now so we don't need to worry about it. Note that `DeepImagePredictor` will continue to use the python implementation due to the dependency on keras.

The python version of resizing uses `imageIO.resizeImage` which uses PIL.